src/routes/transactions/mappers/common/transaction-data.mapper.ts

```diff
@@ -0,0 +1,170 @@
+    @Inject(IContractsRepository)
+    private readonly contractRepository: ContractsRepository,
+    private readonly dataDecodedParamHelper: DataDecodedParamHelper,
+  ) {}
+
+  async mapTransactionData(
+    chainId: string,
+    previewTransactionDto: PreviewTransactionDto,
+    dataDecoded: DataDecoded | null,
+  ): Promise<TransactionData> {
+    const toAddress = await this.addressInfoHelper.getOrDefault(
+      chainId,
+      previewTransactionDto.to,
+    );
+    const isTrustedDelegateCall = await this.isTrustedDelegateCall(
+      chainId,
+      previewTransactionDto.operation,
+      previewTransactionDto.to,
+      dataDecoded,
+    );
+    const addressInfoIndex = await this.buildAddressInfoIndex(
+      chainId,
+      dataDecoded,
+    );
+
+    return new TransactionData(
+      previewTransactionDto.data,
+      dataDecoded,
+      toAddress,
+      previewTransactionDto.value,
+      previewTransactionDto.operation,
+      isTrustedDelegateCall ?? null,
+      addressInfoIndex,
+    );
+  }
+
+  /**
+   * Determines if the transaction is a trusted DELEGATE call.
+   * @param chainId chain id
+   * @param operation transaction operation
+   * @param to transaction target address
+   * @param dataDecoded data decoded to check
+   * @returns null if the transaction operation is not DELEGATE.
+   * If the transaction operation is DELEGATE, and the target {@link Contract}
+   * is trusted, and the {@link DataDecoded} received contains a nested
+   * DELEGATE operation, then true is returned.
+   * Otherwise the function will return false.
+   */
+  async isTrustedDelegateCall(
+    chainId: string,
+    operation: Operation,
+    to: string,
+    dataDecoded: DataDecoded | null,
+  ): Promise<boolean | null> {
+    if (dataDecoded === null || operation !== DELEGATE_OPERATION) return null;
+
+    let contract: Contract;
+    try {
+      contract = await this.contractRepository.getContract(chainId, to);
+    } catch (err) {
+      return false;
+    }
+
+    return (
+      contract.trustedForDelegateCall &&
+      this.dataDecodedParamHelper.hasNestedDelegate(dataDecoded)
+    );
+  }
+
+  /**
+   * Builds a {@link Record<string, AddressInfo>} which contains all the addresses
+   * extracted from {@link DataDecoded} as keys, and their related {@link AddressInfo}
+   * as value.
+   * @param chainId chain id
+   * @param dataDecoded data decoded to use
+   * @returns {@link Record<string, AddressInfo>}
+   */
+  async buildAddressInfoIndex(
+    chainId: string,
+    dataDecoded: DataDecoded | null,
+  ): Promise<Record<string, AddressInfo>> {
+    if (dataDecoded === null || !isArray(dataDecoded.parameters)) return {};
+
+    const addressInfos = await Promise.all(
+      this.getAddressParametersFromDataDecoded(dataDecoded).map((param) =>
+        this.getIfValidAddress(chainId, param),
+      ),
+    );
+
+    return addressInfos
+      .filter((addressInfo) => !isNull(addressInfo))
```

Comment:
Do we need to `filter` here? In the `reduce` function we are also tackling the nullability of `addressInfo` 🤔

Reply:
True, we were double-checking it, thanks. Changed in [9565c0a](https://github.com/5afe/safe-client-gateway-nest/pull/291/commits/9565c0ad8b6c6533745567f36f005f4915859bf4)